### ISSUE TITLE
Transaction pool not clearing all `babe.reportEquivocationUnsigned` transactions

### ISSUE THREAD

**Issue by maxsam4** [paritytech/substrate#8005]
If a validator is equivocating continuously, the nodes keep generating `babe.reportEquivocationUnsigned` transactions. It's expected that these transactions will keep getting mined and the pool will keep emptying. However, currently, a backlog of these transactions starts building up on all non validator nodes. This issue becomes a real problem due to https://github.com/paritytech/substrate/issues/8004

Since an equivocating validator is not getting ignored, it keeps equivocating and nodes keep reporting it. Eventually, every node's tx pool gets full. 

This issue can be reproduced by using something like the following script:
```
const { ApiPromise, Keyring } = require("@polkadot/api");
const path = require('path');
// Path to substrate binary
const binaryPath = path.join(__dirname, 'data', 'substrate');
const exec = require('child_process').exec;
const waitOn = require('wait-on');

const bobStash = "5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc";

async function main() {
    exec(`${binaryPath} --chain local --tmp`);
    const waitOnOpts = {
        resources: [`http://localhost:9933`],
        validateStatus: function (status) {
            return status >= 200;
        },
    };
    await waitOn(waitOnOpts);
    const api = await ApiPromise.create();
    const alice = new Keyring({ type: 'sr25519' }).addFromUri('//Alice');
    await api.tx.sudo.sudo(api.tx.staking.setInvulnerables([])).signAndSend(alice);
    exec(`${binaryPath} --chain local --tmp --alice`);
    
    // Bob will equivocate
    exec(`${binaryPath} --chain local --tmp --bob`);
    exec(`${binaryPath} --chain local --tmp --bob`);
    
    // Adding a bunch more nodes to expedite the issue
    for (let i = 0; i < 50; i++) {
        exec(`${binaryPath} --chain local --tmp`);
    }

    await api.derive.chain.subscribeNewHeads(async (header) => {
        if (header.author?.toString() == bobStash) {
            console.log("Bob the builder still building");
        }
        const pendingTransactions = await api.rpc.author.pendingExtrinsics();
        console.log(`Pending transactions: ${pendingTransactions.length}`);
    });
}

main()
```

I am using 2.0.1 release of substrate.

**Comment by bkchr** [comment]
Hmm, as you already found out, the problem is that the validator is not disabled. You still want that all equivocation reports are being generated and included on chain.

Cc @andresilva

**Comment by maxsam4** [comment]
> You still want that all equivocation reports are being generated and included on chain.

@bkchr The problem here is that the tx queue is building up with duplicate reports. The reports that are valid and get included in the blockchain are fine but you will see that the number of pending reports keep increasing over time and never get mined. Also worth noting that this only happens on non authoring nodes. The queue of authoring nodes remain empty. Perhaps the transactions in the queue are not being propagated to be included?

**Comment by andresilva** [comment]
Currently the automatic misbehavior reports are sent out automatically using unsigned extrinsics. To avoid DDoS these can only be included by block authors. In the example you have provided there is only one authority running and it will not report itself, as such there is no one else to report the equivocations and clean up the queue.

I am not going to argue that the current implementation using unsigned extrinsics is perfect, but your example is an unrealistic scenario. This is expected behavior with the current implementation and something that won't change soon.

**Comment by andresilva** [comment]
@maxsam4 I ran your example code now, here's the output of a run:

```
❯ node queue_bug.js
2021-02-03 15:12:25        API/INIT: RPC methods not decorated: system_unstable_networkState
Pending transactions: 1
Pending transactions: 1
Bob the builder still building
Pending transactions: 1
Bob the builder still building
Pending transactions: 0
Pending transactions: 0
Bob the builder still building
Pending transactions: 1
Bob the builder still building
Pending transactions: 2
Bob the builder still building
Pending transactions: 3
Bob the builder still building
Pending transactions: 4
Bob the builder still building
Pending transactions: 5
Bob the builder still building
Pending transactions: 6
Bob the builder still building
Pending transactions: 7
Bob the builder still building
Pending transactions: 8
Bob the builder still building
Pending transactions: 9
Bob the builder still building
Pending transactions: 10
Bob the builder still building
Pending transactions: 11
Bob the builder still building
Pending transactions: 12
Bob the builder still building
Pending transactions: 13
Bob the builder still building
Pending transactions: 14
Bob the builder still building
Pending transactions: 15
Bob the builder still building
Pending transactions: 16
Pending transactions: 0
Pending transactions: 0
Pending transactions: 0
Bob the builder still building
Pending transactions: 1
Pending transactions: 0
Bob the builder still building
Pending transactions: 1
Bob the builder still building
Pending transactions: 2
Bob the builder still building
Pending transactions: 3
```

So as you can see the transactions are being pruned from the queue once Alice authors a block that includes the equivocation reports. So I think everything is working as expected here, as I said above the current situation may not be perfect due to usage of unsigned extrinsics and the fact that only block authors can include them. 

The situation in this example is exacerbated by the fact that: 1) there is only one other authority that can include the reports, 2) there are 50 nodes running on a single machine.

**Comment by maxsam4** [comment]
@andresilva It seems like you got super lucky that bob got to built 16 blocks in a row in a 2 validator network. Perhaps try running the script a couple more times and for longer? Here's a modified script to lower the peer nodes.

```
const { ApiPromise, Keyring } = require("@polkadot/api");
const path = require('path');
// Path to substrate binary
const binaryPath = path.join(__dirname, 'data', 'substrate');
const exec = require('child_process').exec;
const waitOn = require('wait-on');
const bobStash = "5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc";

async function main() {
    // We'll use this node as our RPC node
    exec(`${binaryPath} --chain local --tmp`);
    const waitOnOpts = {
        resources: [`http://localhost:9933`],
        validateStatus: function (status) {
            return status >= 200;
        },
    };
    await waitOn(waitOnOpts);
    const api = await ApiPromise.create();
    const alice = new Keyring({ type: 'sr25519' }).addFromUri('//Alice');
    await api.tx.sudo.sudo(api.tx.staking.setInvulnerables([])).signAndSend(alice);

    // Alice is a non equivocating validator
    exec(`${binaryPath} --chain local --tmp --alice`);

    // Bob is an equivocating validator
    exec(`${binaryPath} --chain local --tmp --bob`);
    exec(`${binaryPath} --chain local --tmp --bob`);

    // Adding a bunch more nodes to expedite the issue
    for (let i = 0; i < 5; i++) {
        exec(`${binaryPath} --chain local --tmp`);
    }

    await api.derive.chain.subscribeNewHeads(async (header) => {
        if (header.author?.toString() == bobStash) {
            console.log("Bob the builder still building");
        } else {
            console.log("Alice authored a block");
        }
        const pendingTransactions = await api.rpc.author.pendingExtrinsics();
        console.log(`Pending transactions: ${pendingTransactions.length}`);
    });
}

main()
```

Below you will find sample output that shows that tx are not always cleared even when Alice creates a block.
```
$ node equivocate.js 
Alice authored a block
Pending transactions: 1
Alice authored a block
Pending transactions: 0
Bob the builder still building
Pending transactions: 1
Alice authored a block
Pending transactions: 1
Bob the builder still building
Pending transactions: 1
Alice authored a block
Pending transactions: 2
Alice authored a block
Pending transactions: 2
Bob the builder still building
Pending transactions: 2
Bob the builder still building
Pending transactions: 3
Alice authored a block
Pending transactions: 4
Alice authored a block
Pending transactions: 4
Alice authored a block
Pending transactions: 4
Alice authored a block
Pending transactions: 4
Bob the builder still building
Pending transactions: 4
Bob the builder still building
Pending transactions: 5
Bob the builder still building
Pending transactions: 6
Alice authored a block
Pending transactions: 6
Bob the builder still building
Pending transactions: 7
Bob the builder still building
Pending transactions: 7
Alice authored a block
Pending transactions: 8
Alice authored a block
Pending transactions: 8
Bob the builder still building
Pending transactions: 9
Alice authored a block
Pending transactions: 8
Bob the builder still building
Pending transactions: 9
Alice authored a block
Pending transactions: 9
Alice authored a block
Pending transactions: 9
Alice authored a block
Pending transactions: 9
Alice authored a block
Pending transactions: 9
Bob the builder still building
Pending transactions: 10
Bob the builder still building
Pending transactions: 10
Alice authored a block
Pending transactions: 11
Bob the builder still building
Pending transactions: 12
Alice authored a block
Pending transactions: 12
Alice authored a block
Pending transactions: 12
Bob the builder still building
Pending transactions: 13
Bob the builder still building
Pending transactions: 13
Alice authored a block
Pending transactions: 13
Bob the builder still building
Pending transactions: 13
Bob the builder still building
Pending transactions: 15
Alice authored a block
Pending transactions: 13
Bob the builder still building
Pending transactions: 13
Bob the builder still building
Pending transactions: 15
Alice authored a block
Pending transactions: 14
Bob the builder still building
Pending transactions: 15
Bob the builder still building
Pending transactions: 15
```

**Comment by andresilva** [comment]
10 minute run: https://gist.github.com/andresilva/9664ac72045e0fdd087828e72acc39bd

I can conclude that my computer is faster than yours :) I would be worried if the logic for cleaning included transactions was not being triggered or it was incorrect. It seems to be working as expected but as I said this example is an edge case where you have 3 validators running and only 1 is honest. This is an edge case we don't need to support in the real-world.

**Comment by maxsam4** [comment]
Which substrate version are you using? I don't get how "computer speed" has anything to do with this. I am able to reproduce this issue reliably on my 8 core @4.1 GHz machine. The block production is not delayed or anything.

> This is an edge case we don't need to support in the real-world.
This is an MVC, not a use case. This is meant to help you reproduce the issue, this is not the setup I'm rocking in prod.

> I would be worried if the logic for cleaning included transactions was not being triggered or it was incorrect.
This is exactly what I am trying to report with this issue.

I am using Ubuntu 18 FWIW.

**Comment by andresilva** [comment]
> This is exactly what I am trying to report with this issue.

I didn't observe this behavior on my tests hence why I was saying that the test case was not realistic.
I tested with master, going to test with v2.0.1 again now. I didn't mean to be dismissive of your report sorry if it seemed like that.

**Comment by maxsam4** [comment]
Here's the log from almost an hour long on my machine: https://gist.github.com/maxsam4/2a8154ca28a0340eb2fa35732bcdb3ce

I will try to reproduce with substrate master.

I know this is OSS and you don't owe me anything but I was just trying to work towards to a mutually beneficial thing. Sorry if I have been rude, that was definitely not my intention.

**Comment by andresilva** [comment]
I just tested with v2.0.1 and indeed can replicate the issue. A quick search on merged PRs comparing master with v2.0.1 I didn't find anything that was an obvious fix for this. I'll bisect it.

> I know this is OSS and you don't owe me anything but I was just trying to work towards to a mutually beneficial thing. Sorry if I have been rude, that was definitely not my intention.

You were right about the issue and provided a proper test case with code to replicate it. I appreciate this and I'm sorry that I didn't test your exact setup right away. No hard feelings on my side 🤝.

**Comment by andresilva** [comment]
Fixed in master with https://github.com/paritytech/substrate/pull/8041 and https://github.com/paritytech/substrate/pull/8048. Thanks for the report.
